pkgs/development/beam-modules/fetch-mix-deps.nix

```diff
@@ -1,34 +1,49 @@
 { stdenvNoCC, lib, elixir, hex, rebar, rebar3, cacert, git }:
 
-{ name, version, sha256, src, mixEnv ? "prod", debug ? false, meta ? { } }:
-
-stdenvNoCC.mkDerivation ({
-  name = "mix-deps-${name}-${version}";
-
+{ pname
+, version
+, sha256
+, src
+, mixEnv ? "prod"
+, debug ? false
+, meta ? { }
+, ...
+}@attrs:
+
+stdenvNoCC.mkDerivation (attrs // {
   nativeBuildInputs = [ elixir hex cacert git ];
 
-  inherit src;
-
   MIX_ENV = mixEnv;
   MIX_DEBUG = if debug then 1 else 0;
   DEBUG = if debug then 1 else 0; # for rebar3
-
-  configurePhase = ''
+  # the api with `mix local.rebar rebar path` makes a copy of the binary
+  MIX_REBAR = "${rebar}/bin/rebar";
+  MIX_REBAR3 = "${rebar3}/bin/rebar3";
+  # there is a persistent download failure with absinthe 1.6.3
+  # those defaults reduce the failure rate
+  HEX_HTTP_CONCURRENCY = 1;
+  HEX_HTTP_TIMEOUT = 120;
+
+  configurePhase = attrs.configurePhase or ''
+    runHook preConfigure
     export HEX_HOME="$TEMPDIR/.hex";
     export MIX_HOME="$TEMPDIR/.mix";
-    export MIX_DEPS_PATH="$out";
+    export MIX_DEPS_PATH="$TEMPDIR/deps";
 
     # Rebar
-    mix local.rebar rebar "${rebar}/bin/rebar"
-    mix local.rebar rebar3 "${rebar3}/bin/rebar3"
     export REBAR_GLOBAL_CONFIG_DIR="$TMPDIR/rebar3"
     export REBAR_CACHE_DIR="$TMPDIR/rebar3.cache"
+    runHook postConfigure
   '';
 
   dontBuild = true;
 
-  installPhase = ''
+  installPhase = attrs.installPhase or ''
+    runHook preInstall
     mix deps.get --only ${mixEnv}
```

Comment:
Does it make sense to compile the dependencies here with `mix deps.compile` and potentially cache that build for larger (phoenix) projects? Or does elixir/nix not like that?

Reply:
`mix deps.compile` is done in the configure phase of the `mixRelease`.
It could be possible to compile dependencies here.
It did not feel natural to have a compilation step into a step to fetch dependencies.
The other concern I had, was that what if you wanted to compile your dependencies in a specific way.

If you wanted to cache the compilation step, you could potentially add another intermediary step.
I have to say I don't know yet the consequences that it would have.
Perhaps it's something that can be tried for 6 months somewhere so we can make a PR about it later?
Curious if anybody has a more definite opinion.
It could be something to ask on an official elixir forum as well.

Reply:
mix deps.compile doesn't produce deterministic output

Reply:
Oh right, I forgot about the config part of things :thinking: 

bummer :disappointed:

Reply:
@schneebyte thanks for the answer here. I kept on thinking about this.
Do you mean to say that mix deps.compile will produce something different based on the configuration?
I'm just wondering what is producing the non-deterministic behavior, is it some input that we can pass in? Or is it that no matter what we pass in, we will never get deterministic output?

Reply:
Well there is compile time configuration, https://hexdocs.pm/mime/MIME.html for example.
And also the .mix directories in _build/env/lib/xyz/.mix change, but those could be deleted if you pass `--no-deps-check` to next mix command.

I actually got my Elixir project to produce deterministic output, until i added phoenix_live_dashboard to deps, suddenly one .beam file was changing between builds.

If i compile this a few times the beam file changes.
> defmodule Foo do
>   def bar() do
>     ~r/(?<beginning>((.+?\/){3})).*(?<ending>(\/.*){3})/
>   end
> end

Reply:
Thanks a lot for the precision, very interesting!

Reply:
Thank you so much @schneebyte for providing a minimal example! I was actually investigating this a while ago, and it seemed like the built in pcre2 library in Erlang sometimes doesn't produce deterministic compiled regex data. I feel like with the minimal example, we should be able to report a bug to the Erlang project

Reply:
That example was actually wrong, formatting messed it up.
I narrowed it down to the length of the names of named captures?
But only with the sigil in elixir, not sure how to do it with :re.compile.

~r/(?<xxx>.)(?<yyy>.)/
Is fine because xxx and yyy are both length 3

But with
~r/(?<xxxxx>.)(?<yy>.)/
the beam file differs because xxxxx and yy have different lengths.

Weird stuff.